sklearn/linear_model/tests/test_sgd.py

```diff
@@ -1158,12 +1165,77 @@ def test_large_regularization():
     # regularization parameters
     for penalty in ['l2', 'l1', 'elasticnet']:
         model = SGDClassifier(alpha=1e5, learning_rate='constant', eta0=0.1,
-                              n_iter=5, penalty=penalty, shuffle=False)
+                              penalty=penalty, shuffle=False,
+                              tol=-np.inf, max_iter=6)
         with np.errstate(all='raise'):
             model.fit(iris.data, iris.target)
         assert_array_almost_equal(model.coef_, np.zeros_like(model.coef_))
 
 
+def test_tol_parameter():
+    # Test that the tol parameter behaves as expected
+    X = StandardScaler().fit_transform(iris.data)
+    y = iris.target
+
+    # With tol == -np.inf, the number of iteration should be equal to max_iter
+    max_iter = 42
+    model_0 = SGDClassifier(tol=-np.inf, max_iter=max_iter).fit(X, y)
+    assert_equal(max_iter, model_0.n_iter_)
+
+    # With tol > -np.inf, the number of iteration should be less than max_iter
+    max_iter = 2000
+    model_1 = SGDClassifier(tol=0, max_iter=max_iter).fit(X, y)
+    assert_greater(max_iter, model_1.n_iter_)
+
+    # A larger tol should yield a smaller number of iteration
+    model_2 = SGDClassifier(tol=1e-3, max_iter=max_iter).fit(X, y)
+    assert_greater(model_1.n_iter_, model_2.n_iter_)
+
+
+def test_future_and_deprecation_warnings():
+    # Test that warnings are raised. Will be removed in 0.21
+
+    # When all default values are used
+    msg_future = "max_iter and tol parameters have been added in "
+    assert_warns_message(FutureWarning, msg_future, SGDClassifier)
+
+    def init(max_iter=None, tol=None, n_iter=None):
+        SGDClassifier(max_iter=max_iter, tol=tol, n_iter=n_iter)
+
+    # When n_iter is specified
+    msg_deprecation = "n_iter parameter is deprecated"
+    assert_warns_message(DeprecationWarning, msg_deprecation, init, 6, 0, 5)
+
+    # When n_iter=None, and at least one of tol and max_iter is specified
+    assert_no_warnings(init, 100, None, None)
+    assert_no_warnings(init, None, 1e-3, None)
+    assert_no_warnings(init, 100, 1e-3, None)
```

Comment:
Please add assertions for the resulting values of `clf.max_iter` and `clf.tol` for each of these cases.